From 5776ae407bb6f0020c0b0259c18a283b9a7b06fc Mon Sep 17 00:00:00 2001
From: Robert Yang <liezhi.yang@windriver.com>
Date: Fri, 11 Apr 2014 04:15:57 -0400
Subject: [PATCH] pyanaconda: do the image installation based kickstart

Read the required data from the ksdata and do the kickstart installation.

Upstream-Status: Inappropriate [Upstream supports this]

Signed-off-by: Robert Yang <liezhi.yang@windriver.com>
---
 pyanaconda/__init__.py  |    6 +++++-
 pyanaconda/kickstart.py |   23 ++++++++++++-----------
 pyanaconda/livecd.py    |    5 ++++-
 3 files changed, 21 insertions(+), 13 deletions(-)

diff --git a/pyanaconda/__init__.py b/pyanaconda/__init__.py
index 757cf98..b38bd78 100644
--- a/pyanaconda/__init__.py
+++ b/pyanaconda/__init__.py
@@ -94,7 +94,11 @@ class Anaconda(object):
     @property
     def backend(self):
         if not self._backend:
-            b = self.instClass.getBackend()
+            if self.ksdata and self.ksdata.liveimg.liveimg:
+                import pyanaconda.livecd
+                b = pyanaconda.livecd.LiveCDCopyBackend
+            else:
+                b = self.instClass.getBackend()
             self._backend = apply(b, (self, ))
 
         return self._backend
diff --git a/pyanaconda/kickstart.py b/pyanaconda/kickstart.py
index 68b405d..7e87738 100644
--- a/pyanaconda/kickstart.py
+++ b/pyanaconda/kickstart.py
@@ -1506,20 +1506,21 @@ def setSteps(anaconda):
     if ksdata.reboot.action in [KS_REBOOT, KS_SHUTDOWN]:
         dispatch.skip_steps("complete")
 
-    # If the package section included anything, skip group selection.
-    if ksdata.upgrade.upgrade:
-        dispatch.skip_steps("tasksel", "group-selection")
+    if not anaconda.ksdata.liveimg.liveimg:
+        # If the package section included anything, skip group selection.
+        if ksdata.upgrade.upgrade:
+            dispatch.skip_steps("tasksel", "group-selection")
 
-        # Special check for this, since it doesn't make any sense.
-        if packagesSeen:
-            warnings.warn("Ignoring contents of %packages section due to upgrade.")
-    elif havePackages(ksdata.packages):
-        dispatch.skip_steps("tasksel", "group-selection")
-    else:
-        if packagesSeen:
+            # Special check for this, since it doesn't make any sense.
+            if packagesSeen:
+                warnings.warn("Ignoring contents of %packages section due to upgrade.")
+        elif havePackages(ksdata.packages):
             dispatch.skip_steps("tasksel", "group-selection")
         else:
-            dispatch.request_steps("tasksel")
+            if packagesSeen:
+                dispatch.skip_steps("tasksel", "group-selection")
+            else:
+                dispatch.request_steps("tasksel")
 
     if ksdata.ignoredisk.interactive:
         dispatch.request_steps("filtertype", "filter")
diff --git a/pyanaconda/livecd.py b/pyanaconda/livecd.py
index c389c5e..a2290b4 100644
--- a/pyanaconda/livecd.py
+++ b/pyanaconda/livecd.py
@@ -101,7 +101,10 @@ class LiveCDCopyBackend(backend.AnacondaBackend):
         self.supportsPackageSelection = False
         self.skipFormatRoot = True
 
-        self.osimg_path = anaconda.methodstr[9:]
+        if anaconda.ksdata and anaconda.ksdata.liveimg.liveimg:
+            self.osimg_path = anaconda.ksdata.liveimg.liveimg
+        else:
+            self.osimg_path = anaconda.methodstr[9:]
         if not stat.S_ISBLK(os.stat(self.osimg_path)[stat.ST_MODE]):
             anaconda.intf.messageWindow(_("Unable to find image"),
                                _("The given location isn't a valid %s "
-- 
1.7.10.4

